src/domain/session/room/timeline/deserialize.js

```diff
@@ -0,0 +1,285 @@
+    }
+    return new ListBlock(start, nodes);
+}
+
+function parseCodeBlock(options, node) {
+    const { result } = options;
+    let codeNode;
+    for (const child of result.getChildNodes(node)) {
+        codeNode = child;
+        break;
+    }
+    if (!(codeNode && result.getNodeElementName(codeNode) === "CODE")) {
+        return null;
+    }
+    let language = "";
+    const cl = result.getAttributeValue(codeNode, "class") || ""
+    for (const clname of cl.split(" ")) {
+        if (clname.startsWith("language-") && !clname.startsWith("language-_")) {
+            language = clname.substring(9) // "language-".length
+            break;
+        }
+    }
+    return new CodeBlock(language, codeNode.textContent);
+}
+
+function parseImage(options, node) {
+    const { result, mediaRepository } = options;
+    const src = result.getAttributeValue(node, "src") || "";
+    const url = mediaRepository.mxcUrl(src);
+    // We just ignore non-mxc `src` attributes.
+    if (!url) {
+        return null;
+    }
+    const width = result.getAttributeValue(node, "width");
+    const height = result.getAttributeValue(node, "height");
+    const alt = result.getAttributeValue(node, "alt");
+    const title = result.getAttributeValue(node, "title");
+    return new ImagePart(url, { width, height, alt, title });
+}
+
+function buildNodeMap() {
+    let map = {
+        A: { descend: true, parsefn: parseLink },
+        UL: { descend: false, parsefn: parseList },
+        OL: { descend: false, parsefn: parseList },
+        PRE: { descend: false, parsefn: parseCodeBlock },
+        BR: { descend: false, parsefn: () => new NewLinePart() },
+        HR: { descend: false, parsefn: () => new RulePart() },
+        IMG: { descend: false, parsefn: parseImage }
+    }
+    for (const tag of basicNodes) {
+        map[tag] = { descend: true, parsefn: basicWrapper(tag) }
+    }
+    for (let level = 1; level <= 6; level++) {
+        const tag = "h" + level;
+        map[tag] = { descend: true, parsefn: headerWrapper(level) }
+    }
+    return map;
+}
+
+/**
+ * Handlers for various nodes.
+ *
+ * Each handler has two properties: `descend` and `parsefn`.
+ * If `descend` is true, the node's children should be
+ * parsed just like any other node, and fed as a second argument
+ * to `parsefn`. If not, the node's children are either to be ignored
+ * (as in <pre>) or processed specially (as in <ul>).
+ *
+ * The `parsefn` combines a node's data and its children into
+ * an internal representation node.
+ */
+const nodes = buildNodeMap();
+
+function parseNode(options, node) {
+    const { result } = options;
+    if (result.isTextNode(node)) {
+        return new TextPart(result.getNodeText(node));
+    } else if (result.isElementNode(node)) {
+        const f = nodes[result.getNodeElementName(node)];
+        if (!f) {
+            return null;
+        }
+        const children = f.descend ? parseNodes(options, node.childNodes) : null;
```

Comment:
I think this might have slipped the opaque handle approach:
```suggestion
        const children = f.descend ? parseNodes(options, result.getChildNodes(node)) : null;
```

Reply:
:facepalm: